sttp/client/src/main/scala/endpoints/sttp/client/Endpoints.scala

```diff
@@ -0,0 +1,112 @@
+  */
+class Endpoints[R[_]](host: String, val backend: sttp.SttpBackend[R, Nothing]) extends algebra.Endpoints with Urls with Methods {
+
+  /**
+    * A function that, given an `A` and a request model, returns an updated request
+    * containing additional headers
+    */
+  type RequestHeaders[A] = (A, sttp.Request[_, Nothing]) => sttp.Request[_, Nothing]
+
+  /** Does not modify the request */
+  lazy val emptyHeaders: RequestHeaders[Unit] = (_, request) => request
+
+  /**
+    * A function that takes an `A` information and returns a `sttp.Request`
+    */
+  type Request[A] = A => sttp.Request[_, Nothing]
+
+  /**
+    * A function that, given an `A` information and a `sttp.Request`, eventually returns a `sttp.Request`
+    */
+  type RequestEntity[A] = (A, sttp.Request[_, Nothing]) => sttp.Request[_, Nothing]
+
+  lazy val emptyRequest: RequestEntity[Unit] = { case (_, req) => req }
+
+  def request[A, B, C, AB](
+    method: Method, url: Url[A],
+    entity: RequestEntity[B], headers: RequestHeaders[C]
+  )(implicit tuplerAB: Tupler.Aux[A, B, AB], tuplerABC: Tupler[AB, C]): Request[tuplerABC.Out] =
+    (abc: tuplerABC.Out) => {
+      val (ab, c) = tuplerABC.unapply(abc)
+      val (a, b) = tuplerAB.unapply(ab)
+
+      val uri: sttp.Id[sttp.Uri] = sttp.Uri(new URI(s"${host}${url.encode(a)}"))
+      val reqId: sttp.Request[_, Nothing] = method(sttp.sttp.get(uri = uri))
+      entity(b, headers(c, reqId))
+    }
+
+  /**
+    * Trait that indicates how a response should be interpreted
+    */
+  trait SttpResponse[A] {
+    /**
+      * The type of the received body from the server
+      */
+    type RB
+
+    /**
+      * To read the response body
+      */
+    def responseAs: sttp.ResponseAs[RB, Nothing]
+
+    /**
+      * Function to validate the response (headers, code). This can also modify the type of the received body
+      */
+    def validateResponse(response: sttp.Response[RB]): Either[String, A]
+  }
+
+  type Response[A] = SttpResponse[A]
+
+  /** Successfully decodes no information from a response */
+  val emptyResponse: Response[Unit] = new SttpResponse[Unit] {
+    type RB = Unit
+    override def responseAs = sttp.ignore
+    override def validateResponse(response: sttp.Response[Unit]) = {
+      if (response.isSuccess) response.body
+      else Left(s"Unexpected status code: ${response.code}")
+    }
+  }
+
+  /** Successfully decodes string information from a response */
+  val textResponse: Response[String] = new SttpResponse[String] {
+    type RB = String
+    override def responseAs = sttp.asString
+    override def validateResponse(response: sttp.Response[String]) = {
+      if (response.isSuccess) response.body
+      else Left(s"Unexpected status code: ${response.code}")
+    }
+  }
+
+  /**
+    * A function that, given an `A`, eventually attempts to decode the `B` response.
+    */
+  type Endpoint[A, B] = A => R[Either[String, B]]
```

Comment:
I'm not sure about that. The idea, at least for me, behind endpoints was that properly defined protocol should never fail because of expected reasons. And here we say something different. Maybe that's ok, but I would like to understand in which cases we will receive `Left`.

Reply:
According to the [documentation](http://sttp.readthedocs.io/en/latest/responses/basics.html#obtaining-the-response-body):

```
The response.body will be a:
Left(errorMessage) if the request is successful, but response code is not 2xx.
Right(deserializedBody) if the request is successful and the response code is 2xx.
``` 

If you think that defined endpoints should not fail because of expected reasons, `.unsafeBody` may work better in our case

Reply:
Maybe we could leave it up to the user? So have an `apply` which will call `unsafeBody` and some other method that will give back the `Either`. I wouldn't like to force the user to write `right.get` every time even if he is sure that server and client are consistent. Nevertheless, if this comes because of underlying API I leave it up to you to decide :)

Reply:
`unsafeBody` doesn’t sound appealing because it throws.

That being said I agree with @Krever, we tend to give `Endpoint[A, B]` a concrete type that doesn’t require users to deal with failures due to communications. Typically, it is represented by a function `A => F[B]`, where `F` has a `MonadError[F, Exception]` instance. Do we have the same guarantees for `R`, here?

Reply:
(by failures due to communication, I mean network failures but also data marshalling failures)